src/v/kafka/server/group.cc

```diff
@@ -2307,12 +2317,16 @@ group::handle_offset_commit(offset_commit_request&& r) {
 }
 
 ss::future<offset_fetch_response>
-group::handle_offset_fetch(offset_fetch_request&& r) {
+group::handle_offset_fetch(offset_fetch_request&& req) {
+    auto r = std::move(req);
     if (in_state(group_state::dead)) {
-        return ss::make_ready_future<offset_fetch_response>(
-          offset_fetch_response(r.data.topics));
+        co_return r.data.topics;
     }
 
+    // We should see current state for txs. So we need to get write lock before
+    // has_pending_transaction() checks
+    auto lock = co_await _state_lock.hold_write_lock();
```

Comment:
> offset_fetch request can not get transient state for txs stuff
in group service. So we have to get write lock before
has_pending_transaction check.

the locking rules need to be precisely described in the commit message. one thing in particular that needs to be described is why a write lock is held here. the reason that this is important to describe is because the code that executes below taking the write lock and then end of the function where the lock is released does not contain any co_awaits. thus, it already runs without interruption, so what is it protecting?

Reply:
+1. this needs more thought..

I think the issue there is with `_volatile_txns` and `_prepared_txns`.. when we try to mark a txn as prepared the sequence if as follows..

```
volatile_txns.erase(txn);
co_await replicate(prepared_txn);
_prepared_txns.put(txn);
```

So if there is a has_pending_transactions() scheduled before or after replicate(), it sees an inconsistent state (sees no active transactions resulting in incorrect offsets).. so I believe the lock is intended to avoid this problem.. there is a per pid lock too (see with_pid_lock()), so your patch makes it a nested lock which is a code smell.. need more details here.

Reply:
Idea: fetch_offset should see current state of maps inside group service. Because client logic expect it. To see do we have another requests like (commit_tx, abort_tx or smth else)

As you can see another txs requests do some async work. We have 2 different maps for current txs: `_volatile_txs` and `_prepared_txs`
For example:
```
_volatile_txs.erase(tx_it);
...
auto e = co_await _partition->raft()->replicate(...)
...
 _prepared_txs[r.pid] = ptx;
```
group has ongoing transaction, but offset_fetch can not see it. So we wanna be sure that when we ans on this request we do not have ongoing transaction. Also this request is not so frequent. 

How solve this problem:
1) use read_lock in all txs request.
2) use write lock for offset_fetch request. So after this we can be sure that we do not have anoter onging txs request and state is correct.


Will add it in comments

Reply:
Thinking out loud.. can we do this without these extra locks?

We already have a per pid lock that prevents interleaving of operations for a single pid. All these state changing operations are called within that mutex scope.

```
with_pid_lock(req.pid, ... return prepare_txn(req))
```

Now all we want is a consistent view of _volatile_txs and _prepared_txs, so we just need to make those updates in an atomic block.  Something like..

```
...
 auto e = co_await _partition->raft()->replicate(...);

---- atomic ----
_volatile_txs.erase(tx_it); ...  // handle error cases, cleanup from _volatile in all exception cases
_prepared_txs[r.pid] = ptx;
-------------------

.....
```

So we write the prepare (or abort or whatever) marker to the log first and then update all the in-memory state together. Its just that the entry stays in `_volatile_txs` a tiny bit longer than before but `has_pending_txns()` shouldn't see any inconsistency. Does that not work for some reason that I'm not seeing?

(As the code is already complex, just trying to avoid nested locks if there is no need for it, correct me if I'm wrong).

Reply:
In your suggestion I see one problem. If replication will fail we will have situation where txs is still inside group_service. But client got error for commit. So we will have inconsistent state. Is it bad - idk. Need to think about it

Looks like we still need sync only prepare_tx and fetch_offset request. Looks like we can delete read_lock from another tx requests.

Reply:
Had a brief chat with @VadimPlh about the lock free approach, don't think there is any consistency issue with the error handling bit. We just need to cleanup the _volatile_txns map in all error cases. @rystsov may want to take a second look.

Reply:
Ok, its seems fine to go without looks if we put `_volatile_txs.erase(tx_it);` after `_prepared_txs[r.pid] = ptx;` only put  a comment that the order is important for `handle_offset_fetch` to be correct. Also don't forget to recapture the iterator because it maybe invalidated by a concurrent transactions changing `_volatile_txs`.